File: backend/src/routes/alerts.js
// @ts-check

import { Router } from 'express';
import { normalize } from '../services/normalizer.js';
import { getConfig, getConfigInt } from '../services/config.js';

const router = Router();

/** @typedef {{ [key: string]: string | number }} QueryParams */

/**
 * @param {unknown} rawValue
 * @param {string} fieldName
 * @returns {{ ok: true, value: number } | { ok: false, error: string }}
 */
function parseNonNegativeInteger(rawValue, fieldName) {
  const value = Number(rawValue);
  if (!Number.isInteger(value) || value < 0) {
    return { ok: false, error: `${fieldName} must be a non-negative integer` };
  }
  return { ok: true, value };
}

/**
 * POST /api/alerts/ingest
 * 多源事件接入：接收原始日志，规范化后存入数据库
 *
 * Body: { source: 'waf'|'nids'|'hids'|'pikachu'|'soc', events: [...] }
 */
router.post('/ingest', (/** @type {any} */ req, /** @type {any} */ res) => {
  const { source, events } = req.body;
  if (!source || !Array.isArray(events)) {
    return res.status(400).json({ error: 'source and events[] required' });
  }

  // 从 system_config 动态读取
  const validSources = getConfig('ingest.valid_sources', 'waf,nids,hids,pikachu,soc').split(',').map((s) => s.trim());
  const maxBatchSize = getConfigInt('ingest.max_batch_size', 1000);

  if (!validSources.includes(source)) {
    return res.status(400).json({ error: `source must be one of: ${validSources.join(', ')}` });
  }
  if (events.length === 0) {
    return res.status(400).json({ error: 'events[] must not be empty' });
  }
  if (events.length > maxBatchSize) {
    return res.status(400).json({ error: `events[] exceeds max batch size (${maxBatchSize})` });
  }

  const db = req.db;
  const insertRaw = db.prepare(
    'INSERT INTO raw_events (source, raw_json) VALUES (?, ?)'
  );
  const findSimilar = db.prepare(`
    SELECT id FROM alerts
    WHERE source = @source AND title = @title AND severity = @severity
      AND created_at >= datetime('now', '-' || @window_hours || ' hours')
    ORDER BY created_at DESC LIMIT 1
  `);
  const incrementCount = db.prepare(
    'UPDATE alerts SET event_count = event_count + 1 WHERE id = ?'
  );
  const insertAlert = db.prepare(`
    INSERT INTO alerts (source, severity, title, description, src_ip, dst_ip, mitre_tactic, mitre_technique, raw_event_id, event_count)
    VALUES (@source, @severity, @title, @description, @src_ip, @dst_ip, @mitre_tactic, @mitre_technique, @raw_event_id, 1)
  `);

  const clusteringWindow = getConfigInt('ingest.clustering_window_hours', 1);

  /** @type {Record<string, unknown>[]} */
  const results = [];
  const batchInsert = db.transaction(() => {
    for (const event of events) {
      const rawResult = insertRaw.run(source, JSON.stringify(event));
      const normalized = normalize(source, event);
      normalized.raw_event_id = rawResult.lastInsertRowid;

      // 聚簇：同 source + title + severity 在窗口内的归并
      const existing = findSimilar.get({ ...normalized, window_hours: clusteringWindow });
      if (existing) {
        incrementCount.run(existing.id);
        results.push({ id: existing.id, clustered: true, ...normalized });
      } else {
        const alertResult = insertAlert.run(normalized);
        results.push({ id: Number(alertResult.lastInsertRowid), ...normalized });
      }
    }
  });

  try {
    batchInsert();
  } catch (err) {
    console.error('Ingest batch failed:', err);
    return res.status(500).json({ error: 'Failed to ingest events' });
  }
  res.json({ ingested: results.length, alerts: results });
});

/**
 * GET /api/alerts
 * 查询告警列表，支持筛选和分页
 */
router.get('/', (/** @type {any} */ req, /** @type {any} */ res) => {
  const { status, severity, source, limit = 50, offset = 0 } = req.query;
  const db = req.db;

  const parsedLimit = parseNonNegativeInteger(limit, 'limit');
  if (!parsedLimit.ok) return res.status(400).json({ error: parsedLimit.error });
  const parsedOffset = parseNonNegativeInteger(offset, 'offset');
  if (!parsedOffset.ok) return res.status(400).json({ error: parsedOffset.error });

  /** @type {string[]} */
  const where = [];
  /** @type {QueryParams} */
  const params = {};
  if (status) { where.push('status = @status'); params.status = String(status); }
  if (severity) { where.push('severity = @severity'); params.severity = String(severity); }
  if (source) { where.push('source = @source'); params.source = String(source); }

  const whereClause = where.length ? `WHERE ${where.join(' AND ')}` : '';
  const sql = `SELECT * FROM alerts ${whereClause} ORDER BY created_at DESC LIMIT @limit OFFSET @offset`;
  params.limit = parsedLimit.value;
  params.offset = parsedOffset.value;

  const alerts = db.prepare(sql).all(params);
  const countParams = Object.fromEntries(
    Object.entries(params).filter(([key]) => !['limit', 'offset'].includes(key))
  );
  const total = db.prepare(`SELECT COUNT(*) as count FROM alerts ${whereClause}`).get(countParams);

  res.json({ total: total.count, alerts });
});

/**
 * GET /api/alerts/:id
 * 查询单条告警详情
 */
router.get('/:id', (/** @type {any} */ req, /** @type {any} */ res) => {
  const alert = req.db.prepare(
    'SELECT a.*, r.raw_json, r.source as raw_source FROM alerts a LEFT JOIN raw_events r ON a.raw_event_id = r.id WHERE a.id = ?'
  ).get(req.params.id);
  if (!alert) {
    return res.status(404).json({ error: 'Alert not found' });
  }
  if (alert.raw_json) {
    try {
      alert.raw_json = JSON.parse(alert.raw_json);
    } catch { /* keep as string if not valid JSON */ }
  }
  res.json(alert);
});

/**
 * Valid alert state transitions: open → analyzing → analyzed → resolved
 * @type {Record<string, string[]>}
 */
const VALID_TRANSITIONS = {
  open: ['analyzing'],
  analyzing: ['analyzed', 'open'],     // allow rollback to open on failure
  analyzed: ['resolved', 'open'],      // allow reopen
  resolved: ['open'],                  // allow reopen
};

/**
 * PATCH /api/alerts/:id/status
 * 更新告警状态（人工反馈闭环），校验状态转换合法性
 */
router.patch('/:id/status', (/** @type {any} */ req, /** @type {any} */ res) => {
  const { status } = req.body;
  if (!['open', 'analyzing', 'analyzed', 'resolved'].includes(status)) {
    return res.status(400).json({ error: 'Invalid status' });
  }

  const db = req.db;
  const current = db.prepare('SELECT id, status FROM alerts WHERE id = ?').get(req.params.id);
  if (!current) {
    return res.status(404).json({ error: 'Alert not found' });
  }

  const allowed = VALID_TRANSITIONS[current.status] || [];
  if (!allowed.includes(status)) {
    return res.status(400).json({
      error: `Invalid transition: ${current.status} → ${status}. Allowed: ${allowed.join(', ') || 'none'}`,
    });
  }

  db.prepare('UPDATE alerts SET status = ? WHERE id = ?').run(status, req.params.id);

  // Write audit log for status change
  db.prepare(
    `INSERT INTO audit_logs (trace_id, alert_id, event_type, data) VALUES (?, ?, ?, ?)`
  ).run(
    `manual-${Date.now()}`,
    String(req.params.id),
    'alert_status_change',
    JSON.stringify({ from: current.status, to: status })
  );

  res.json({ updated: true });
});

export default router;

File: agent/routers/alerts.py
"""
Alerts router — POST /ingest, GET /, GET /:id, PATCH /:id/status

Migrated from Express backend/src/routes/alerts.js
"""

from __future__ import annotations

import json
import logging
import sqlite3
from typing import Any, Optional

from fastapi import APIRouter, HTTPException, Query, Path
from pydantic import BaseModel, Field

from agent.routers._db import get_db
from collector.normalizer import normalize as _normalize_raw

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


# ---------------------------------------------------------------------------
# Normalizer bridge — delegates to collector.normalizer and returns a dict
# ---------------------------------------------------------------------------

def normalize(source: str, raw_event: Any) -> dict:
    """Normalize a raw event dict via the canonical collector normalizer.

    Returns a plain dict with the keys expected by the alerts table.
    """
    event = raw_event if isinstance(raw_event, dict) else {}
    normalized_alert = _normalize_raw(source, event)
    d = normalized_alert.to_dict()
    # Drop fields not needed for DB insertion (keep the subset used by ingest)
    return {
        "source": d.get("source"),
        "severity": d.get("severity"),
        "title": d.get("title"),
        "description": d.get("description"),
        "src_ip": d.get("src_ip"),
        "dst_ip": d.get("dst_ip"),
        "mitre_tactic": d.get("mitre_tactic"),
        "mitre_technique": d.get("mitre_technique"),
    }


# ---------------------------------------------------------------------------
# Config helpers (from shared _helpers module)
# ---------------------------------------------------------------------------

from agent.routers._helpers import get_config as _get_config, get_config_int as _get_config_int


# ---------------------------------------------------------------------------
# Pydantic models
# ---------------------------------------------------------------------------

class IngestRequest(BaseModel):
    source: str
    events: list[dict[str, Any]]


class IngestResponse(BaseModel):
    ingested: int
    alerts: list[dict[str, Any]]


class AlertListResponse(BaseModel):
    total: int
    alerts: list[dict[str, Any]]


class StatusUpdateRequest(BaseModel):
    status: str


VALID_TRANSITIONS: dict[str, list[str]] = {
    "open": ["analyzing"],
    "analyzing": ["analyzed", "open"],
    "analyzed": ["resolved", "open"],
    "resolved": ["open"],
}


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@router.post("/ingest", response_model=IngestResponse)
async def ingest_events(request: IngestRequest):
    """Multi-source event ingestion: normalize raw events and store as alerts."""
    conn = get_db()
    try:
        source = request.source
        events = request.events

        valid_sources_str = _get_config(conn, "ingest.valid_sources", "waf,nids,hids,pikachu,soc")
        valid_sources = [s.strip() for s in valid_sources_str.split(",")]
        max_batch = _get_config_int(conn, "ingest.max_batch_size", 1000)

        if source not in valid_sources:
            raise HTTPException(status_code=400, detail=f"source must be one of: {', '.join(valid_sources)}")
        if len(events) == 0:
            raise HTTPException(status_code=400, detail="events[] must not be empty")
        if len(events) > max_batch:
            raise HTTPException(status_code=400, detail=f"events[] exceeds max batch size ({max_batch})")

        clustering_window = _get_config_int(conn, "ingest.clustering_window_hours", 1)
        results: list[dict[str, Any]] = []

        for event in events:
            # Insert raw event
            cursor = conn.execute(
                "INSERT INTO raw_events (source, raw_json) VALUES (?, ?)",
                (source, json.dumps(event)),
            )
            raw_event_id = cursor.lastrowid

            normalized = normalize(source, event)
            normalized["raw_event_id"] = raw_event_id

            # Clustering: merge if same source + title + severity within window
            existing = conn.execute(
                """
                SELECT id FROM alerts
                WHERE source = ? AND title = ? AND severity = ?
                  AND created_at >= datetime('now', '-' || ? || ' hours')
                ORDER BY created_at DESC LIMIT 1
                """,
                (normalized["source"], normalized["title"], normalized["severity"], str(clustering_window)),
            ).fetchone()

            if existing:
                conn.execute("UPDATE alerts SET event_count = event_count + 1 WHERE id = ?", (existing["id"],))
                results.append({"id": existing["id"], "clustered": True, **normalized})
            else:
                cursor = conn.execute(
                    """
                    INSERT INTO alerts (source, severity, title, description, src_ip, dst_ip,
                                        mitre_tactic, mitre_technique, raw_event_id, event_count)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                    """,
                    (
                        normalized["source"], normalized["severity"], normalized["title"],
                        normalized["description"], normalized.get("src_ip"), normalized.get("dst_ip"),
                        normalized.get("mitre_tactic"), normalized.get("mitre_technique"),
                        raw_event_id,
                    ),
                )
                results.append({"id": cursor.lastrowid, **normalized})

        conn.commit()
        return IngestResponse(ingested=len(results), alerts=results)
    except HTTPException:
        raise
    except Exception as exc:
        logger.exception("Ingest batch failed")
        raise HTTPException(status_code=500, detail="Failed to ingest events") from exc


@router.get("/", response_model=AlertListResponse)
async def list_alerts(
    status: Optional[str] = None,
    severity: Optional[str] = None,
    source: Optional[str] = None,
    limit: int = Query(default=50, ge=0),
    offset: int = Query(default=0, ge=0),
):
    """Query alerts list with filtering and pagination."""
    conn = get_db()
    where_parts: list[str] = []
    params: dict[str, Any] = {}
    if status:
        where_parts.append("status = :status")
        params["status"] = status
    if severity:
        where_parts.append("severity = :severity")
        params["severity"] = severity
    if source:
        where_parts.append("source = :source")
        params["source"] = source

    where_clause = f"WHERE {' AND '.join(where_parts)}" if where_parts else ""
    params["limit"] = limit
    params["offset"] = offset

    alerts = conn.execute(
        f"SELECT * FROM alerts {where_clause} ORDER BY created_at DESC LIMIT :limit OFFSET :offset",
        params,
    ).fetchall()

    count_params = {k: v for k, v in params.items() if k not in ("limit", "offset")}
    total_row = conn.execute(
        f"SELECT COUNT(*) as count FROM alerts {where_clause}", count_params
    ).fetchone()

    return AlertListResponse(
        total=total_row["count"],
        alerts=[dict(a) for a in alerts],
    )


@router.get("/{alert_id}")
async def get_alert(alert_id: int = Path(...)):
    """Query single alert details."""
    conn = get_db()
    alert = conn.execute(
        """SELECT a.*, r.raw_json, r.source as raw_source
           FROM alerts a LEFT JOIN raw_events r ON a.raw_event_id = r.id
           WHERE a.id = ?""",
        (alert_id,),
    ).fetchone()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    result = dict(alert)
    if result.get("raw_json"):
        try:
            result["raw_json"] = json.loads(result["raw_json"])
        except (json.JSONDecodeError, TypeError):
            pass
    return result


@router.patch("/{alert_id}/status")
async def update_alert_status(alert_id: int, body: StatusUpdateRequest):
    """Update alert status with state transition validation."""
    status = body.status
    if status not in ("open", "analyzing", "analyzed", "resolved"):
        raise HTTPException(status_code=400, detail="Invalid status")

    conn = get_db()
    current = conn.execute("SELECT id, status FROM alerts WHERE id = ?", (alert_id,)).fetchone()
    if not current:
        raise HTTPException(status_code=404, detail="Alert not found")

    allowed = VALID_TRANSITIONS.get(current["status"], [])
    if status not in allowed:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid transition: {current['status']} -> {status}. Allowed: {', '.join(allowed) or 'none'}",
        )

    conn.execute("UPDATE alerts SET status = ? WHERE id = ?", (status, alert_id))

    import time
    conn.execute(
        "INSERT INTO audit_logs (trace_id, alert_id, event_type, data) VALUES (?, ?, ?, ?)",
        (f"manual-{int(time.time() * 1000)}", str(alert_id), "alert_status_change",
         json.dumps({"from": current["status"], "to": status})),
    )
    conn.commit()
    return {"updated": True}
